### ISSUE TITLE
Support cards with multiple targets.

### ISSUE THREAD

**Issue by bilbo2** [magarena/magarena#1526]
When looking at unimplemented cards and grouping them by status, the second biggest group (after the needs-groovy, which just needs some effort to write a specific groovy code, but there is no lack of feature support from  the game engine) is "multiple-targets" with 590 occurrences.

It would be nice to implement the possibility of selecting multiple targets for a card, as quite a lot of useful cards have more than one target (like Symbiosis) and some mechanics are thus also completely unsupported, like Proliferate (any number of targets), damage division (Forked Lightning deals 4 damage divided amongst 1, 2 or 3 targets) or fighting (target creature fights another target creature). Quite often is the actual number of targets limited to two.

I am not sure what exactly blocks support for multiple targets, so before trying to add the support in some kind of merge request, I'd like to gather some feedback.

Those are the issues I am currently aware of:
- need for more complex GUI dialogs for some kinds of targeting
- in AI, multiple targets will mean much more choices to explore, which could mean effectively weaker/slower AI if many multi-target spells will be used.

Is there more caveats when implementing this feature?

---

**Comment by melvinzhang** [comment]
The biggest block may the existing code base which assumes single target, for example currently one `MagicEvent` can have up to one target, thus it has methods like `getTargetChoice`, `processTarget`, etc, and `MagicEvent` is used in many places.

One idea I've had for multiple targets is to create several events, one event per target. We already have places like cost payment which consists of multiple events, so this approach seems to work better with the current code base. However, this approach has drawbacks if information from one target affects what happens to the next target, since it is not easy to pass information between events.

Rather than starting on the code, it will probably be a good idea to discuss the design/approach of adding multiple targets in this issue first.

**Comment by melvinzhang** [comment]
Thinking about this further, it looks like modifying MagicEvent to support multiple targets should be possible as well as long as we can keep the legacy single target methods to only use the first target. Then add multiple target methods for events that deal with them.

There is also the issue of target pickers, this is something used by the AIs. For example, MMAB uses it to reduce the tree width and MCTS uses it during simulations.

**Comment by bilbo2** [comment]
That was the intention to gather some feedback before starting any coding here...

In some cards the targets are very related, i.e. Hunt Down (Target creature blocks target creature this turn if able.), sometimes they need just to be different (Toils of Night and Day - You may tap or untap target permanent, then you may tap or untap another target permanent) and sometimes they are completely unrelated (Trap Essence - Counter target creature spell. Put two +1/+1 counters on up to one target creature.). The last case might be a bit easier, as it could be perhaps done as two separate effects internally.

I see a lot of Magic[Something]TargetPicker ... I guess most of them are specialized to some kind of spell/effect, so I guess the new 2+target pickers would need to be tailored to specific kinds of effects too ...

I see quite a lot of arbitrary numbers (some decks would have big benefit for example from trample and little from first strike and vice versa) and some possibly bad decisions from them. For example if creature does have first strike, "upgrading" it with doublestrike would be (often) beneficial, but here it has score of 0 (i.e. seen as useless). I could perhaps look there and check them later :) Might be worth trying to use evolutionary algorithms to tweak them ...

**Comment by bilbo2** [comment]
As for the AI ... maybe inspire a bit with AlphaGo .... in their first version they let it learn from a huge database of games played by humans. AlphaGo learned how to play well, but also learned some of the human mistakes and some of the workarounds for human limitations in brain capacity, concentration, etc ... that are not necessary to follow, as the AI have its own (different) limits.

Later they started it again, with only rudimentary knowledge of basic rules (learn from zero) and let it learn by playing against itself. It needed some time to get from random play to be decent, but in the end it ended up being better than the one trained from human games.

Here the filters are also human-generated (with opinionated values) and ultimately we might want to throw them away and replace with AI-learned rules :)

But when comparing with go, MtG is harder in several aspects:
- go has stable rules for thousands of years. MtG is evolving with every new set release and subset of MtG actually supported by the game is also evolving, perhaps even more frequently ... so we are chasing a moving target.
- go has 2 types of stones and 19x19 board with complete information. MtG has tens of thousands of distinct cards, decks can contain almost anything (limited by only few deck building rules) and some information is hidden (opponent's hand, library and face-down creatures), so there comes guessing and bluffing... On the board, number or permanents, mana or basically just anything is not limited....

So maybe AlphaGo-like approach would work, but the computational cost of training a decent AI may be a bit prohibitive ....

**Comment by melvinzhang** [comment]
> The last case might be a bit easier, as it could be perhaps done as two separate effects internally.

Agree, a good starting point is to find the smallest thing we could support.

> I guess the new 2+target pickers would need to be tailored to specific kinds of effects too

If we focus on the independent effects first, we can reuse the existing pickers.

> I see quite a lot of arbitrary numbers

That was one of the motivation for the MCTS AI. It should not need any of these numbers, though in practice it still uses these as some of the option generation uses scores to filter out the bad cases. For example, blocking options generation uses these numbers to decide which blocks are profitable. Would be nice to either get rid of them or determine them in a systemic way.

**Comment by firemind** [comment]
> So maybe AlphaGo-like approach would work

I've gone down this rabbit hole before :smile: 

I believe the main challenge here is encoding the board state in a way that is palatable by a DNN which could then be trained as a value network using the win/loss ratios from MCTS. Preferably this encoding would not treat each card as a separate entity but use its properties as inputs so as to treat functional reprints the same. But since any zone can contain an arbitrary amount of cards and their order does not matter, there is no clean mapping of the game state to model inputs as there is in Go. At least I've not yet been able to find one.

Currently I'm trying to use DNNs to solve the simpler partial problems of Mulligans and combat math respectively. Because guiding MCTS in those areas might improve AI performance without having to encode the entire game state.
